### PR TITLE
cleaning typing CI and config

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

After source unification, it is time to revisit how we run typing...

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @carmocca @akihironitta @borda @justusschock @awaelchli

### CODE REVIEW

File: requirements/typing.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+mypy==0.982
+torch==1.13.1
```
Comment (carmocca) [inline requirements/typing.txt:2]: Can you split this PR into 2?

One where you fix the paths and CI config
And another where you upgrade the torch version

File: .github/workflows/code-checks.yml [hunk 2]
```diff
@@ -34,36 +34,17 @@ jobs:
...
-        pip install torch==1.12 --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
-        python ./requirements/pytorch/adjust-versions.py requirements/pytorch/extra.txt
         # todo: adjust requirements for both code-bases
-        pip install \
-          -r requirements/fabric/base.txt \
-          -r requirements/pytorch/base.txt \
-          -r requirements/pytorch/extra.txt \
-          -r requirements/app/base.txt \
-          -r requirements/app/cloud.txt \
-          -r requirements/app/ui.txt \
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:59]: We need to install all these requirements still. Each of our dependencies might have their own typing definitions which we want our mypy workflow to use when checking

**Review by carmocca — CHANGES_REQUESTED** [review]

File: requirements/typing.txt (hunk 1 above)
Comment (Borda) [reply requirements/typing.txt:2]:
```suggestion
torch==1.12.0
```

File: .github/workflows/code-checks.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/code-checks.yml:59]: So far I understood, instaling `types-*` is full replacement
but regarding these all `*/base.txt` are installed with this package so the remaining are 
- pytorch/extra.txt
- app/cloud.txt
- app/ui.txt

Comment (carmocca) [reply .github/workflows/code-checks.yml:59]: The types change looks good.

Then those 3 should be kept

File: .github/workflows/code-checks.yml [hunk 3]
```diff
@@ -34,36 +34,17 @@ jobs:
...
-        restore-keys: pip-
-        enableCrossOsArchive: true
+        path: .mypy_cache
+        key: mypy-${{ hashFiles('requirements/typing.txt') }}
 
     - name: Install dependencies
       run: |
-        pip install mypy==0.982  # TODO: pull this verion deom requirements/test.txt
-        pip install torch==1.12 --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
-        python ./requirements/pytorch/adjust-versions.py requirements/pytorch/extra.txt
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:51]: We should still adjust versions with the proper torch version (1.12 in this case)

Comment (Borda) [reply .github/workflows/code-checks.yml:51]: The adjustment is "only" for `torchvision` and that is used in examples so not really needed for typing checks :)

File: .github/workflows/code-checks.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/code-checks.yml:59]: addressed in [f4d1167](https://github.com/Lightning-AI/lightning/pull/16860/commits/f4d11675d1047e8f02931694d23298cee35f6fa8)

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by lantiga — APPROVED** [review]

**Pull request merged**